File: Subset_sum_problem.c
#include<stdio.h>
#include<stdlib.h>
// int static Arr[1000][1000];
int subsetSumProblem(int arr[],int sum, int n){
    int count=0;
    if(sum=0){
        return 1;
    }
    if(n==0 && sum==0){
        return 1;
    }
    if(n==0 && sum!=0){
        return 0;
    }
    if(arr[n-1]<=sum){
        count = count + (subsetSumProblem(arr,sum-arr[n-1],n-1)+subsetSumProblem(arr,sum,n-1));
        printf(" c1 %d",count);
        return count;
    }
    else{
        count = count + subsetSumProblem(arr,sum,n-1);
        printf(" c2 %d",count);
        return count;
    }
}

int main(){
    int arr[4]={1,5,7,1};
    int sum=6;
    int value = subsetSumProblem(arr,sum,4);
    // memset(Arr,0,)
    // printf("%d",value);
    return 0;
}




// Not solved

File: SubsetSumProblem.java
package Array;
import java.util.*;
public class SubsetSumProblem {
    static boolean isSumPossible(int arr[], int n, int sum){
        boolean t[][] = new boolean[n+1][sum+1];
        for(int i=0;i<=n;i++){
            if(sum==0){
                t[i][0] = true;
            }
        }
        for(int i=1;i<=sum;i++){
            if(n==0){
                t[0][i] =false;
            }
        }
        
        for(int i=1;i<=n;i++){
            for(int j=1;j<=sum;j++){
                if(arr[i-1]>sum){
                    t[i][j] = t[i-1][j];
                }

                else{
                    t[i][j] = (t[i-1][j] || t[i-1][j-arr[i-1]]);
                }
            }
        }
        return t[n][sum];
    }
    public static void main(String[] args){
        int n;
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        int [] arr = new int [n];
        for(int i=0;i<n;i++){
            arr[i] = sc.nextInt();
        }
        int sum;
        sum = sc.nextInt();
        boolean result = isSumPossible(arr, n, sum);
        System.out.println(result);
        sc.close();
    }
}
